src/networkapi.jl

```diff
@@ -1481,5 +1481,158 @@ function validate(rs::ReactionSystem, info::String = "")
     validated
 end
 
+"""
+    iscomplexbalanced(rs::ReactionSystem, rates::Vector)
+
+Constructively compute whether a network will have complex-balanced equilibrium
+solutions, following the method in [this paper](https://link.springer.com/article/10.1007/s10910-015-0498-2#Sec3). Accepts a map of rates [k1 => 1.0, k2 => 2.0,...]k2
+"""
+
+using Combinatorics, LinearAlgebra
+
+function iscomplexbalanced(rs::ReactionSystem, rates::Dict{Any, Float64}) 
+    if length(rates) != numparams(rs) 
+        error("The number of reaction rates must be equal to the number of parameters")
+    end
+
+    sm = speciesmap(rs)
+    cm = reactioncomplexmap(rs)
+    complexes, D = reactioncomplexes(rs)
+    rxns = reactions(rs)
+    nc = length(complexes); nr = numreactions(rs); nm = numspecies(rs)
+
+    if !all(r->ismassaction(r, rs), rxns) error("Not mass action") end
+
+    # Construct kinetic matrix, K
+    K = zeros(nr, nc) 
+    for c in 1:nc
+        complex = complexes[c]
+        for (r, dir) in cm[complex]
+            rxn = rxns[r]
+            if dir == -1
+                K[r, c] = rates[rxn.rate]
+            end
+        end
+    end
+
+    L = -D*K
+    S = netstoichmat(rs)
+
+    # Compute ρ using the matrix-tree theorem
+    g = incidencematgraph(rs); R = ratematrix(rs, rates)
+    ρ = matrixtree(g, R)
+    @assert isapprox(L*ρ, zeros(nc), atol=1e-12) 
+
+    # Determine if 1) ρ is positive and 2) D^T Ln ρ lies in the image of S^T
+    if all(x -> x > 0, ρ)
+        img = D'*log.(ρ)
+        if rank(S') == rank(hcat(S', img)) return true else return false end 
+    else
+        return false
+    end
+end
+
+# """
+#     rateweightedgraph(rs::ReactionSystem, rates::Vector)
+# 
+# Generate an annotated reaction complex graph of a reaction system, where the nodes are annotated with the reaction complex they correspond to and the edges are annotated with the reaction they correspond to and the rate of the reaction. 
+# """
+# 
+# function rateweightedgraph(rs::ReactionSystem, rates::Dict{Any, Float64}) 
+#     if length(rates) != numparams(rs) 
+#         error("The number of reaction rates must be equal to the number of parameters")
+#     end
+# 
+#     complexes, D = reactioncomplexes(rs)
+#     rxns = reactions(rs)
+# 
+#     g = incidencematgraph(rs)
+#     rwg = MetaDiGraph(g)
+# 
+#     for v in vertices(rwg)
+#         set_prop!(rwg, v, :complex, complexes[v])
+#     end
+# 
+#     for (i, e) in collect(enumerate(edges(rwg)))
+#         rxn = rxns[i]
+#         set_prop!(rwg, Graphs.src(e), Graphs.dst(e), :reaction, rxn)
+#         set_prop!(rwg, Graphs.src(e), Graphs.dst(e), :rate, rates[rxn.rate])
+#     end
+# 
+#     rwg
+# end
+
+function ratematrix(rs::ReactionSystem, rates::Dict{Any, Float64}) 
+    if length(rates) != numparams(rs) 
+        error("The number of reaction rates must be equal to the number of parameters")
+    end
```

Comment:
```suggestion
```
See comments above about how to numerically calculate the rates (I'd suggest validating and calculating them above, and then just passing the final numeric rate vector into this function. It may help speedup this function by passing in a more concrete type for `rates`, i.e. one that doesn't depend on `Any`).